### ISSUE TITLE
Issues with reduce_by_segment with zip_iterators

### ISSUE THREAD

**Issue by abagusetty** [oneapi-src/oneDPL#207]
Hi,

I was trying to figure out a test case that involves using `dpl::reduce_by_segment` with `zip_iterators(tuple)` and was facing some difficultly with compiling it.

Can someone please comment if there is something wrong the way the test case is setup or otherwise.

```cpp
#define PSTL_USE_PARALLEL_POLICIES 0
#define _GLIBCXX_USE_TBB_PAR_BACKEND 0

#include <CL/sycl.hpp>
#include <oneapi/dpl/execution>
#include <oneapi/dpl/algorithm>
#include <oneapi/dpl/iterator>
#include <oneapi/dpl/functional>

#include <functional>
#include <iostream>
#include <vector>

int main()
{
    sycl::queue q(sycl::gpu_selector{});

    std::vector<int> keys1{11, 11, 21, 20, 21, 21, 21, 37, 37};
    std::vector<int> keys2{11, 11, 20, 20, 20, 21, 21, 37, 37};
    std::vector<int> values{0, 1, 2, 3, 4, 5, 6, 7, 8};
    std::vector<int> output_keys1(keys1.size());
    std::vector<int> output_keys2(keys2.size());    
    std::vector<int> output_values(values.size());

    int* d_keys1         = sycl::malloc_device<int>(9, q);
    int* d_keys2         = sycl::malloc_device<int>(9, q);
    int* d_values        = sycl::malloc_device<int>(9, q);
    int* d_output_keys1  = sycl::malloc_device<int>(9, q);
    int* d_output_keys2  = sycl::malloc_device<int>(9, q);
    int* d_output_values = sycl::malloc_device<int>(9, q);

    q.memcpy(d_keys1, keys1.data(), sizeof(int)*9);
    q.memcpy(d_keys2, keys2.data(), sizeof(int)*9);
    q.memcpy(d_values, values.data(), sizeof(int)*9);

    auto begin_keys_in = oneapi::dpl::make_zip_iterator(d_keys1, d_keys2);
    auto end_keys_in   = oneapi::dpl::make_zip_iterator(d_keys1 + 9, d_keys2 + 9);
    auto begin_keys_out= oneapi::dpl::make_zip_iterator(d_output_keys1, d_output_keys2);

    auto new_last = oneapi::dpl::reduce_by_segment(oneapi::dpl::execution::make_device_policy(q),
						   begin_keys_in, end_keys_in, d_values, begin_keys_out, d_output_values);

    q.memcpy(output_keys1.data(), d_output_keys1, sizeof(int)*9);
    q.memcpy(output_keys2.data(), d_output_keys2, sizeof(int)*9);    
    q.memcpy(output_values.data(), d_output_values, sizeof(int)*9);
    q.wait();

    // Expected output
    // {11, 11}: 1
    // {21, 20}: 2
    // {20, 20}: 3
    // {21, 20}: 4
    // {21, 21}: 11
    // {37, 37}: 15
    for(int i=0; i<9; i++) {
      std::cout << "{" << output_keys1[i] << ", " << output_keys2 << "}: " << output_values[i] << std::endl;
    }
}
```
**Environment**
    Target device and vendor: Intel GPUs
    DPC++ version: Intel(R) oneAPI DPC++/C++ Compiler 2021.2.0 (2021.x.0.20210323)

**Comment by jiejanezhang** [comment]
Thank you for reporting. The team is investigating it.

**Comment by MikeDvorskiy** [comment]
I've investigated the case. Currently the iterator-based version of the `reduce_by_segment `doesn't support such case with zip_iterator.
In general, instead of the fancy iterators I would recommend to use "views", as more modern functional approach.

BTW, I'm working on a range-based version of `reduce_by_key` algorithm. A working draft implementation is available by link [(https://github.com/oneapi-src/oneDPL/pull/234)]. 
Also, please have a look at the mentioned above example, re-written with range API: 
(https://github.com/oneapi-src/oneDPL/pull/234/files#diff-553a55ce79907da0f3372464f4a44ce8e1153c96c3d6acf08e28c329b70bb37fR19)

```
int32_t
main()
{
    sycl::queue q(sycl::default_selector{});

    const int n = 9, n_res = 5;
    int a[n] =  {11, 11, 21, 20, 21, 21, 21, 37, 37};
    int a1[n] = {11, 11, 20, 20, 20, 21, 21, 37, 37};
    int b[n] = {0,  1,  2,  3,  4,  5,  6,  7,  8};

    sycl::buffer<int> A(a, sycl::range<1>(n));
    sycl::buffer<int> A1(a1, sycl::range<1>(n));
    sycl::buffer<int> B(b, sycl::range<1>(n));
    sycl::buffer<int> C(n);           // output keys
    sycl::buffer<int> C1(n);           // output keys
    sycl::buffer<int> D(n);           // output values

    using namespace oneapi::dpl::experimental::ranges;

    auto res = reduce_by_key(oneapi::dpl::execution::make_device_policy(q),
        zip_view(views::all_read(A), views::all_read(A1)), 
        views::all_read(B), zip_view(views::all_write(C), views::all_write(C1)), views::all_write(D));

    auto res_keys1 = views::host_all(C);
    auto res_keys2 = views::host_all(C1);
    auto res_values = views::host_all(D);

    ::std::cout << "Output: " << std::endl;
    for(int i = 0; i <  res; ++i)
        ::std::cout << "{" << res_keys1[i] << ", " << res_keys2[i] << "}: " << res_values[i] << std::endl;
    ::std::cout << ::std::endl;
}
```

**Comment by abagusetty** [comment]
@MikeDvorskiy Can you please comment on the status of /pull/234

**Comment by akukanov** [comment]
Hello @abagusetty,

There are two work-in-progress patches for this issue; one is #234 that adds a range-based version of reduce_by_segment, and another is #285 that attempts to fix the core issue with zip_iterator in a generic way. Both are expected to be finished in July, in worst case August.

**Comment by abagusetty** [comment]
> Hello @abagusetty,
> 
> There are two work-in-progress patches for this issue; one is #234 that adds a range-based version of reduce_by_segment, and another is #285 that attempts to fix the core issue with zip_iterator in a generic way. Both are expected to be finished in July, in worst case August.

Thank so much @akukanov for the update. It would be interesting to checkout the functionality with the fancy_iterators.

**Comment by abagusetty** [comment]
After #285 merged, I have tried compiling the above example (with USM ptrs) and seems like there are still couple of errors. I was wondering if someone can take a peek at this:

```
In file included from onedpl.cpp:5:
In file included from /opt/intel/oneapi/dpl/2021.4.0/linux/include/oneapi/dpl/execution:52:
In file included from /opt/intel/oneapi/dpl/2021.4.0/linux/include/oneapi/dpl/pstl/algorithm_impl.h:31:
In file included from /opt/intel/oneapi/dpl/2021.4.0/linux/include/oneapi/dpl/internal/../pstl/parallel_backend.h:22:
In file included from /opt/intel/oneapi/dpl/2021.4.0/linux/include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:27:
In file included from /opt/intel/oneapi/dpl/2021.4.0/linux/include/oneapi/dpl/internal/../pstl/iterator_impl.h:24:
/opt/intel/oneapi/dpl/2021.4.0/linux/include/oneapi/dpl/pstl/tuple_impl.h:415:14: error: no viable overloaded '='
        next = other.next;
        ~~~~ ^ ~~~~~~~~~~


/opt/intel/oneapi/dpl/2021.4.0/linux/include/oneapi/dpl/pstl/tuple_impl.h:428:5: note: candidate template ignored: could not match 'std::tuple' against 'oneapi::dpl::__internal::tuple'
    operator=(const ::std::tuple<U1, U...>& other)
    ^
/opt/intel/oneapi/dpl/2021.4.0/linux/include/oneapi/dpl/pstl/tuple_impl.h:414:22: error: no viable overloaded '='
        holder.value = other.holder.value;
        ~~~~~~~~~~~~ ^ ~~~~~~~~~~~~~~~~~~
```
Compiler: Intel(R) oneAPI DPC++/C++ Compiler 2021.3.0 (2021.3.0.20210619)
oneDPL (main): 3ea3cfe25b88624e252e141f43c4daa78cd50a4a
